### ISSUE TITLE
Why running the program built with bazel to train the model is so slow?

### ISSUE THREAD

**Issue by Kongsea** [tensorflow/serving#146]
I have written a program to train a CNN model using TensorFlow, which trained the model very quickly and the output can be seen in the following:

> Initialized!
> Step 0 (epoch 0.00), 88.4 ms
> Minibatch loss: 2.505, learning rate: 0.010000
> Minibatch error: 47.7%
> Validation error: 48.2%
> Step 18 (epoch 0.95), 27.6 ms
> Minibatch loss: 2.200, learning rate: 0.006302
> Minibatch error: 43.0%
> Validation error: 36.2%
> Step 36 (epoch 1.91), 22.3 ms
> Minibatch loss: 2.174, learning rate: 0.003774
> Minibatch error: 33.6%
> Validation error: 33.6%
> Test error: 42.0%

However, after I insert the TensorFlow serving code to export the model and build the program with "bazel build", the program is running very slowly when I run the generated program of bazel build. The output of the running procedure is shown in the following:

> Initialized!
> Step 0 (epoch 0.00), 303.7 ms
> Minibatch loss: 2.505, learning rate: 0.010000
> Minibatch error: 47.7%
> Validation error: 48.2%
> Step 18 (epoch 0.95), 5409.5 ms
> Minibatch loss: 2.201, learning rate: 0.006302
> Minibatch error: 43.8%
> Validation error: 36.5%
> Step 36 (epoch 1.91), 5138.0 ms
> Minibatch loss: 2.173, learning rate: 0.003774
> Minibatch error: 30.5%
> Validation error: 34.2%
> Exporting trained model to /home/kong/400G/LUNA_MM/modelExport...
> Done exporting!
> Test error: 40.6%

As you can see, the time spent is many times longer than the original program (without TensorFlow serving to export the model). So my question is, since the original program and the built program are almost the same, why the program built with bazel is running so slowly? Thank for your kind answers and suggestions.

**Comment by brummell** [comment]
I'm having the same problem, and it's essentially stalled project development for us. Any chance we could get some potential guidance with this? Even places to start if we're debugging ourselves?

**Comment by Kongsea** [comment]
Now I have to train the model independently and then write a new program just to load the pre-trained model and export it using the TensorFlow serving instead of training it with TF serving which costs enormous time. 

Does anyone have a better way? Thank you.

**Comment by sukritiramesh** [comment]
Hi @Kongsea, a couple of clarifications:
- for the first version (without TF serving export), are intermediate checkpoints saved as well?
- for the second version, is export invoked only at the very end?

Could you provide code snippets for both cases? Will help with debugging. Thanks!

**Comment by Kongsea** [comment]
@sukritiramesh Sorry to reply late.

The checkpoints are only saved after the training process and the export is actually only in the end.

My code is as following which is based on the MNIST example of TensorFlow, and you can have a look.

```
for step in xrange(int(num_epochs * train_size) // BATCH_SIZE):
    offset = (step * BATCH_SIZE) % (train_size - BATCH_SIZE)
    batch_data = train_data[offset:(offset + BATCH_SIZE), ...]
    batch_labels = train_labels[offset:(offset + BATCH_SIZE)]
    batch_data = batch_data.reshape([-1, IMAGE_SIZE*IMAGE_SIZE*NUM_CHANNELS])
    feed_dict = {train_data_node: batch_data,train_labels_node: batch_labels}

    _, l, lr, predictions = sess.run([optimizer, loss, learning_rate, train_prediction],
                                            feed_dict=feed_dict)

    EVAL_FREQUENCY = train_size//BATCH_SIZE
    if step % EVAL_FREQUENCY == 0:
        elapsed_time = time.time() - start_time
        start_time = time.time()
        print('Step %d (epoch %.2f), %.1f ms' %
                      (step, float(step) * BATCH_SIZE / train_size,
                      1000 * elapsed_time / EVAL_FREQUENCY))
        print('Minibatch loss: %.3f, learning rate: %.6f' % (l, lr))
        trE = error_rate(predictions, batch_labels)
        print('Minibatch error: %.1f%%' % trE)
        valE = error_rate(eval_in_batches(validation_data, sess), validation_labels)
        print('Validation error: %.1f%%' % valE)
else:
    if not EXPORT:
        checkpoint_path = (WORK_DIRECTORY+'/Mtest'+'/model.ckpt')
        saver.save(sess, checkpoint_path, global_step=step)
    else:
        export_path = os.path.join(WORK_DIRECTORY, 'modelExport')
        print('Exporting trained model to {}...'.format(export_path))
        model_exporter = exporter.Exporter(saver)
        signature = exporter.classification_signature(input_tensor=train_data_node, scores_tensor=train_prediction)
        model_exporter.init(sess.graph.as_graph_def(),default_graph_signature=signature)
        model_exporter.export(export_path, tf.constant(1), sess)
        print('Done exporting!')
```

**Comment by pingfaichan** [comment]
Build with the -c opt --copt=-mavx options may help.
e.g.
bazel build -c opt --copt=-mavx //tensorflow_serving/example:mnist_inference_2

**Comment by gautamvasudevan** [comment]
Closing due to staleness. If this is still an issue, please file a new updated issue with current steps to reproduce the bug. If this is a question, please ask it on: 

https://stackoverflow.com/questions/tagged/tensorflow-serving

Thanks!
